src/prefect/utilities/tasks.py

```diff
@@ -8,13 +10,170 @@
+    """
+    Map a function that adds tasks to a flow elementwise across one or more
+    tasks.  Arguments that should _not_ be mapped over should be wrapped with
+    `prefect.unmapped`.
+
+    This can be useful when wanting to create complicated mapped pipelines
+    (e.g. ones using control flow components like `case`).
+
+    Args:
+        - func (Callable): a function that adds tasks to a flow
+        - *args: task arguments to map over
+        - flow (Flow, optional): The flow to use, defaults to the current flow
+            in context if no flow is specified. If specified, `func` must accept
+            a `flow` keyword argument.
+        - **kwargs: keyword task arguments to map over
+
+    Returns:
+        - Any: the output of `func`, if any
+
+    Example:
+
+    ```python
+    from prefect import task, case, apply_map
+    from prefect.tasks.control_flow import merge
+
+    @task
+    def inc(x):
+        return x + 1
+
+    @task
+    def is_even(x):
+        return x % 2 == 0
+
+    def inc_if_even(x):
+        with case(is_even(x), True):
+            x2 = inc(x)
+        return merge(x, x2)
+
+    with Flow("example") as flow:
+        apply_map(inc_if_even, range(10))
+    """
+    from prefect.tasks.core.constants import Constant
+
+    no_flow_provided = flow is None
+    if no_flow_provided:
+        flow = prefect.context.get("flow", None)
+        if flow is None:
+            raise ValueError("Couldn't infer a flow in the current context")
+    assert isinstance(flow, prefect.Flow)  # appease mypy
+
+    # Check if args/kwargs are valid first
+    for x in itertools.chain(args, kwargs.values()):
+        if not isinstance(x, (prefect.Task, unmapped, Sequence)):
+            raise TypeError(
+                f"Cannot map over non-sequence object of type `{type(x).__name__}`"
+            )
+
+    flow2 = prefect.Flow("temporary flow")
+    # A mapping of all the input args -> (is_mapped, is_constant)
+    arg_info = {}
+    # A mapping of the ids of all constants -> the Constant task.
+    # Used to convert constants to constant tasks if needed
+    id_to_const = {}
+
+    def preprocess(a: Any) -> "prefect.Task":
+        a2 = as_task(a, flow=flow2)
+        is_mapped = not isinstance(a, unmapped)
+        is_constant = isinstance(a2, Constant)
+        arg_info[a2] = (is_mapped, is_constant)
+        if not is_constant:
+            flow.add_task(a2)  # type: ignore
+        if is_mapped and is_constant:
+            id_to_const[id(a2.value)] = a2  # type: ignore
+        return a2
+
+    args2 = [preprocess(a) for a in args]
+    kwargs2 = {k: preprocess(v) for k, v in kwargs.items()}
+
+    # Construct a temporary flow for the subgraph
+    with prefect.context(mapped=True):
+        with flow2:
+            if no_flow_provided:
+                res = func(*args2, **kwargs2)
+            else:
+                res = func(*args2, flow=flow2, **kwargs2)
```

Comment:
See the docstring on `flow` here: https://github.com/PrefectHQ/prefect/blob/3e919c399e7cf5baa149cc6cdf52371504493f41/src/prefect/utilities/tasks.py#L42-L44

If `flow` is provided as a parameter, we assume the user is using the imperative api inside `func` (and that there's no active `flow` in context). This feels intuitive to me at least, since prefect has a convention in other places of passing in `flow` if one isn't/can't be inferred.

Reply:
Yea that all makes sense to me, but this code block is passing a `Flow` to the user-provided `func` which doesn't seem necessary; whether the user is using the functional or imperative API, your `with flow2:` takes care of putting the relevant Flow object in context, so you could have:
```python
with flow2:
    res = func(*args2, **kwargs2)
```

To be more concrete, this errors out:
```python
from prefect import task, Flow, apply_map

@task
def random_task(y):
    return y + 1

def mapper(x):
    z = random_task(x)
    q = random_task(z)

f = Flow("example")

apply_map(mapper, range(10), flow=f)
# TypeError: mapper() got an unexpected keyword argument 'flow'
```

Reply:
Right, but that's documented in the docstring to `apply_map`.

> If specified, `func` must accept a `flow` keyword argument. 

My main goal here was to make this easy to use with both the functional and imperative api. If a user wants to map a function like:

```python
def func(x, flow):
    flow.add_task(foo)
    flow.add_edge(upstream_task=bar, downstream_task=foo)

apply_map(func, x)
```

should they grab the flow from the context instead, or wrap it with a lambda/partial to ensure the `flow` is passed in? I can see arguments for either behavior - the current design (if you pass in a `flow`, it's forwarded to the `func`) to me doesn't feel *wrong*, just different than what you're expecting. Happy to change, just want to clarify why things are the way they are.

Reply:
Ahhhh ok ok I missed that in the docstring, I understand now.  Yea I think that's a great callout, as there are some users that live entirely within the imperative API.   I'm 👍 on this now, thanks for explaining